ldp/nn/handlers/transformer_handler.py

```diff
@@ -0,0 +1,1007 @@
+from functools import cache, partial, wraps
+from pathlib import Path
+from typing import Any, Concatenate, ParamSpec, Self, TypeVar, cast
+
+import accelerate
+import torch
+import torch.distributed as dist
+import tree
+from dask import config
+from dask.distributed import Client
+from dask_cuda import LocalCUDACluster
+from dask_jobqueue import SLURMCluster
+from pydantic import BaseModel, ConfigDict, Field, field_validator
+from torch import nn
+from torch.cuda import nccl
+from torch.distributed.fsdp import (
+    BackwardPrefetch,
+    FullStateDictConfig,
+    FullyShardedDataParallel,
+    MixedPrecision,
+    ShardingStrategy,
+    StateDictType,
+)
+from torch.nn.functional import pad
+from transformers import PreTrainedModel, PreTrainedTokenizer
+from transformers.generation.utils import GenerateDecoderOnlyOutput
+from transformers.tokenization_utils_base import BatchEncoding
+
+from ldp.graph.async_torch import AsyncBufferedWorker, AsyncTorchModule
+
+from ..lm_config import LMConfig, TorchDType  # noqa: TID252
+from ..utils import REPO_ROOT, set_seed  # noqa: TID252
+from .chunking import TensorChunker
+from .module_handler import ModuleExecutionInterface, ModuleHandler
+
+if sys.version_info >= (3, 12):
+    from typing import overload
+else:
+    from typing_extensions import overload  # noqa: UP035
+
+
+config.set({
+    # We have no use for rebooting workers in aviary for now, and rebooting workers
+    # is annoying when debugging.
+    "distributed.scheduler.allowed-failures": 0,
+    # FSDP forward/backward passes can take way longer than the default warning at 3s
+    "distributed.admin.tick.limit": "30s",
+    # Gives us more time to debug a downed worker. TODO: see if there are negative consequences
+    # of having this always enabled
+    "distributed.comm.timeouts.connect": "300s",
+    "distributed.comm.timeouts.tcp": "300s",
+})
+
+logger = logging.getLogger(__name__)
+
+TReturn = TypeVar("TReturn")
+TParams = ParamSpec("TParams")
+
+
+def is_conversation(messages) -> bool:
+    """Check if messages is an instance of Conversation."""
+    return isinstance(messages, list) and all(
+        isinstance(msg, dict)
+        and all(
+            isinstance(key, str) and isinstance(value, str)
+            for key, value in msg.items()
+        )
+        for msg in messages
+    )
+
+
+def get_unused_port() -> int:
+    """Find an unused port by creating a temporary socket."""
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+        s.bind(("", 0))
+        return s.getsockname()[1]
+
+
+class ExecutionMode(StrEnum):
+    LOCAL_MACHINE = auto()
+    SLURM_CLUSTER = auto()
+
+
+class FSDPConfig(BaseModel):
+    model_config = ConfigDict(extra="forbid")
+
+    offload_cpu: bool = Field(
+        default=False,
+        description="Whether to offload model parameters and gradients to CPU.",
+    )
+    activation_checkpointing: bool = False
+    cpu_ram_efficient_loading: bool = False
```

Comment:
Do you mind adding a `description` to this one, stating what it integrates with? Also, feel free to punt on this one

Reply:
I don't want to overload our code with docstrings - it makes the code less readable. These are all FSDP arguments that are well-documented in the FSDP docs.